lacci/lib/shoes/constants.rb

```diff
@@ -26,4 +27,18 @@ def self.find_lib_dir
     HALF_PI = 1.57079632679489661923
     PI = 3.14159265358979323846
   end
+
+  changelog_content = File.read("CHANGELOG.md")
+
+  release_name_pattern = /^## \[(\d+\.\d+\.\d+)\] - (\d{4}-\d{2}-\d{2})$/m
+
+  release_matches = changelog_content.scan(release_name_pattern)
+
+  latest_release = release_matches.max_by { |version, _date| Gem::Version.new(version) }
+
+  if latest_release
+    RELEASE_NAME, RELEASE_BUILD_DATE = latest_release
+  else
+    puts "most likely something wrong in constants.rb file. or changelog"
```

Comment:
If there's an easy way to use log_init and @log for this, it's a good idea. But this is mergeable as-is.

Reply:
hmm

 how to init_log..? when i just do it individually its throwing no instance error(as by default instance is nil)
 

and when i try to set instance , its giving me too many instances error.

 what i  dng wrong?


```
# frozen_string_literal: true

require_relative "log"
require "scarpe/components/modular_logger"

module Shoes
  class Changelog
    include Shoes::Log

    def initialize
      Shoes::Log.instance = Scarpe::Components::ModularLogImpl.new
      log_init("Changelog")
      puts "Changelog initialized"
    end

    def get_latest_release_info
      changelog_content = File.read("CHANGELOG.md")

      release_name_pattern = /^## \[(\d+\.\d+\.\d+)\] - (\d{4}-\d{2}-\d{2})$/m

      release_matches = changelog_content.scan(release_name_pattern)

      latest_release = release_matches.max_by { |version, _date| Gem::Version.new(version) }

      if latest_release
        @log.debug("Found release #{latest_release[0]} in CHANGELOG.md")
        { RELEASE_NAME: latest_release[0], RELEASE_BUILD_DATE: latest_release[1] }
      else
        { RELEASE_NAME: nil, RELEASE_BUILD_DATE: nil }
      end
    end
  end
end
```

usage:

```
  changelog_instance = Shoes::Changelog.new
  RELEASE_INFO = changelog_instance.get_latest_release_info
  RELEASE_NAME = RELEASE_INFO[:RELEASE_NAME]
  RELEASE_BUILD_DATE = RELEASE_INFO[:RELEASE_BUILD_DATE]
```

Reply:
For starters, you shouldn't create a new instance of ModularLogImpl. You'll only set that if running *outside* of Scarpe, which sets up logging with a specific config, etc.

Reply:
You just log_init() during initialize, then use @log.debug, @log.warn, etc.

Reply:
> You just log_init() during initialize, then use @log.debug, @log.warn, etc.

this is resulting in instance being nil error

```
./exe/scarpe examples/button_alert.rb --dev  
/home/pavan/Desktop/scarpe/lacci/lib/shoes/log.rb:46:in `log_init': undefined method `logger_for_component' for nil:NilClass (NoMethodError)

      @log = Shoes::Log.instance.logger_for_component(component)
                                ^^^^^^^^^^^^^^^^^^^^^
```
that's the reason i tried to create a new instance.

Reply:
Ah, got it. The problem is that you're trying to log really early, before Shoes is really even initialized and before Scarpe has been included. So you'd need to wait until the log instance had been initialized.

You could do that with a const_missing. But definitely easier to just use puts instead of log for now.

I'm trying to think if const_missing is the right answer, or if there's some other event you should be able to catch...

Reply:
You're getting the instance missing if you try to log during startup. You're getting the duplicate instance if you try to create one too early, so that Scarpe will get the duplicate error when it tries (since yours was the first.)

Reply:
hm ok so as of now i am leaving this at puts and seperated changelog logic to seperate file so we can do it if its needed

Reply:
Works for me. At some point I should go through and figure out how we want to do logging at startup here, but in the mean time that would work great.